GVFS/GVFS.Common/Enlistment.cs

```diff
@@ -53,6 +55,7 @@ public abstract class Enlistment
 
         public string EnlistmentRoot { get; }
         public string WorkingDirectoryRoot { get; }
+        public string StorageRoot { get; }

```

Comment:
This name is ambiguous. What are we storing? I was initially going to recommend `GitStorageRoot`, but I see later that we add the `.git` path to the working directory to get the `.git` folder. Whatever name we use here needs to contrast with things like the shared object cache.

Perhaps `LocalStorageRoot`? That can give some information that this is information limited to the enlistment.

Or, do you plan to put anything other than the `.git` folder in this location? Why not put the `.git` folder inside `EnlistmentRoot` and then point `DotGitRoot` there?

We have an opportunity here to think carefully about all of our options here, and later we will be stuck with our decisions. I just want to make sure we've considered all options.

Reply:
To add a bit of context, our libprojfs filesystem works much like any other stackable filesystem on Linux.  We've got a [design document](https://github.com/github/libprojfs/blob/master/docs/design.md#stackable-file-systems) which gets into some of the details as well as a our long-term plans; for now, we're using FUSE and a user-space implementation to thrash out issues which turn up getting VFSForGit to run -- exactly like this one!  :-)

One difference from a filter-based approach, like the PrjFlt NTFS filter or the Mac kauth filter, is that **everything** actually lives in the lower storage directory, not just `.git/` but the entire contents of the working directory (minus those files and directories which have not yet been hydrated).  If a user does `touch src/foo`, the new `foo` file is actually created as `.gvfs/lower/foo`, and is also visible as `src/foo`.  And you can technically read the contents of that file at either location, or modify it in either place, etc.

But the difference is that if you directly modify files in `.gvfs/lower/`, our libprojfs filesystem doesn't see those actions, and sends no events to the VFSForGit provider.  Moreover, if you try to read a not-yet-hydrated file in `.gvfs/lower/`, you just see an empty file; nothing happens to hydrate it.  You only get the hydration and projection actions if you work in `src/`.

This is a pretty typical thing on Linux, familiar to anyone using OverlayFS or eCryptFS.  But it does introduce a whole second "lower storage" directory.  And we've learned that we need to point the GVFS provider at that directory to avoid stalls when it goes looking at the contents of `.git/` before the initial directory is hydrated.

I'd be fine with `LocalStorageRoot` or whatever you think best in terms of naming.  To answer your two specific questions:
> Or, do you plan to put anything other than the .git folder in this location?

Yes, everything visible in the `src/` folder (i.e., the entire working directory's contents) actually is stored in `.gvfs/lower`, including the `.git/` folder.

> Why not put the .git folder inside EnlistmentRoot and then point DotGitRoot there?

I think, off the top of my head, that would break the general functionality of the `src/` folder, which is where the Git working directory is expected to be.  I don't think VFSForGit in general would accommodate such a change, but I might be mistaken!

Reply:
Thanks for the details! Since you are having a local copy of every hydrated file _and_ the `.git` folder in this directory, I think `LocalStorageRoot` is a good enough descriptor. It would not be appropriate to only update `DotGitRoot`.